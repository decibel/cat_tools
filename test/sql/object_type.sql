\set ECHO none

\i test/setup.sql

-- test_role is set in test/deps.sql

SET LOCAL ROLE :use_role;
CREATE FUNCTION pg_temp.major()
RETURNS int LANGUAGE sql IMMUTABLE AS $$
SELECT current_setting('server_version_num')::int/100
$$;

CREATE FUNCTION pg_temp.extra_types()
RETURNS text[] LANGUAGE sql IMMUTABLE AS $$
SELECT '{}'::text[]
  || CASE WHEN pg_temp.major() < 905 THEN '{policy,transform}'::text[] END
  || CASE WHEN pg_temp.major() < 903 THEN '{event trigger}'::text[] END
$$;

CREATE TEMP VIEW obj_type AS
  SELECT object_type::text COLLATE "C", true AS is_real
    FROM cat_tools.enum_range_srf('cat_tools.object_type') r(object_type)
  UNION ALL -- INTENTIONALLY UNION ALL! We want dupes if something gets hosed
  SELECT u COLLATE "C", false
    FROM unnest(pg_temp.extra_types()) u
  ORDER BY 1 -- Intentionally done by ordinal
;

SELECT plan(
  0
  + 4 -- no_use tests
  + 1 -- definition
  + 3 * (SELECT count(*)::int FROM obj_type)

  + 3 -- object__regtype_catalog

  -- Catalog search path
  + 1
  + 3 * 2

  + 1 -- objects__shared
);

SET LOCAL ROLE :no_use_role;
SELECT throws_ok(
  format( 'SELECT NULL::%I', typename )
  , '42704' -- undefined_object; not exactly correct, but close enough
  , NULL
  , 'Permission denied trying to use types'
)
  FROM (VALUES
    ('cat_tools.constraint_type')
    , ('cat_tools.procedure_type')
    , ('cat_tools.object_type')
  ) v(typename)
;
SELECT throws_ok(
  format( 'SELECT cat_tools.object__catalog( NULL::%I )', argtype )
  , '42501' -- insufficient_privilege
  , NULL
  , 'Permission denied trying to run functions'
)
  FROM (VALUES
    ('text'::regtype)
  ) v(argtype)
;

SET LOCAL ROLE :use_role;

SELECT function_returns(
  'cat_tools'
  , 'object__catalog'
  , array[ 'cat_tools.object_type'::name ]
  , 'regclass'::name
);

-- object__regtype_catalog
SELECT throws_ok(
  $$SELECT cat_tools.object__reg_type_catalog('int'::regtype)$$
  , '42809'
  , 'integer is not a object identifier type'
  , $$Invalid object identifier type throws correct error$$
);
CREATE TYPE pg_temp.regtest AS RANGE(subtype = int);
SELECT throws_ok(
  $$SELECT cat_tools.object__reg_type_catalog('pg_temp.regtest'::regtype)$$
  , '0A000'
  , 'object identifier type regtest is not supported'
  , $$Invalid pseudotype throws correct error$$
);

SELECT is(
  cat_tools.object__reg_type_catalog('regclass'::regtype)
  , 'pg_class'::regclass
  , $$Sanity-check cat_tools.object__reg_type_catalog('regclass'::regtype)$$
); -- TODO: all reg*

/*
 * It doesn't seem worth it to hand-check all of these. Just make sure we get a valid relation for all of them.
 *
 * NOTE: Since object__catalog returns regclass we know anything it returns must at least be a valid relation.
 * Should probably at least verify that whatever is returned lives in pg_catalog...
 */
SELECT lives_ok(
      CASE WHEN is_real THEN format( $$SELECT * FROM cat_tools.object__catalog(%L)$$, object_type )
      ELSE 'SELECT 1'
      END
      , format( $$lives_ok: SELECT * FROM cat_tools.object__catalog(%L)$$, object_type )
    )
  FROM obj_type
;

SELECT lives_ok(
      CASE WHEN is_real THEN format( $$SELECT * FROM cat_tools.object__reg_type(%L)$$, object_type )
      ELSE 'SELECT 1'
      END
      , format( $$lives_ok: SELECT * FROM cat_tools.object__reg_type(%L)$$, object_type )
    )
  FROM obj_type
;


-- Verify object__address_classid
SELECT is(
      cat_tools.object__address_classid(object_type)
      , CASE
          WHEN object_type::text LIKE '% column' THEN 'pg_class'::regclass
          ELSE cat_tools.object__catalog(object_type)
        END
      , format( 'Verify cat_tools.object__address_classid(%L)', object_type )
    )
  FROM obj_type
;

-- Verify catalog correctness
SELECT lives_ok(
  $$SET search_path = public,tap,pg_catalog$$ -- Note that we must explicitly set pg_catalog at the end of the path
  , 'Change search_path'
);

SELECT lives_ok(
  'CREATE TABLE pg_class()'
  , 'Create bogus pg_class table'
);
SELECT lives_ok(
  'CREATE TYPE regclass AS ENUM()'
  , 'Create bogus regclass type'
);
SELECT isnt(
  'pg_class'::pg_catalog.regclass
  , 'pg_catalog.pg_class'::pg_catalog.regclass
  , $$Simple 'pg_class'::pg_catalog.regclass should not return pg_catalog.pg_class$$
);
SELECT isnt(
  'regclass'::regtype
  , 'pg_catalog.regclass'::regtype
  , $$Simple 'regclass'::regtype should not return pg_catalog.regtype$$
);
SELECT is(
  cat_tools.object__catalog('table')
  , 'pg_catalog.pg_class'::pg_catalog.regclass
  , $$cat_tools.object__catalog('table') returns pg_catalog.pg_class$$
);
SELECT is(
  cat_tools.object__reg_type('table')
  , 'pg_catalog.regclass'::regtype
  , $$cat_tools.object__catalog('table') returns pg_catalog.pg_class$$
);

SELECT bag_eq(
  $$SELECT * FROM cat_tools.objects__shared_srf()$$
  , array(
    SELECT type::cat_tools.object_type FROM
      -- types and their catalogs
      (SELECT *, cat_tools.object__catalog(type) FROM cat_tools.enum_range_srf('cat_tools.object_type') type) a
      JOIN pg_catalog.pg_class c ON c.oid = a.object__catalog
      WHERE NOT EXISTS(
          -- Does the table have a column with relid or namespace in the column name?
          SELECT 1
            FROM information_schema.columns
            WHERE table_schema='pg_catalog'
              AND table_name = a.object__catalog::text
              AND column_name ~ 'relid|namespace'
        )
        AND c.reltablespace=(SELECT oid FROM pg_tablespace WHERE spcname = 'pg_global')
  )
  , 'Verify objects__shared_src() returns correct values'
);

\i test/pgxntool/finish.sql

--select name,setting from pg_settings where name ~ '^lc_';

-- vi: expandtab ts=2 sw=2
