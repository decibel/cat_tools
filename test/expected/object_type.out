\set ECHO none
1..222
ok 1 - sanity check size of pg_temp.extra_types()
ok 2 - sanity check size of pg_temp.obj_type
ok 3 - Permission denied trying to use types
ok 4 - Permission denied trying to use types
ok 5 - Permission denied trying to use types
ok 6 - Permission denied trying to run functions
ok 7 - Function cat_tools.object__catalog(cat_tools.object_type) should return regclass
ok 8 - Invalid object identifier type throws correct error
ok 9 - Invalid pseudotype throws correct error
ok 10 - Sanity-check cat_tools.object__reg_type_catalog('regclass'::regtype)
ok 11 - check addressability for object type 'access method'
ok 12 - check addressability for object type 'aggregate'
ok 13 - check addressability for object type 'cast'
ok 14 - check addressability for object type 'collation'
ok 15 - check addressability for object type 'composite type'
ok 16 - check addressability for object type 'composite type column'
ok 17 - check addressability for object type 'conversion'
ok 18 - check addressability for object type 'database'
ok 19 - check addressability for object type 'default acl'
ok 20 - check addressability for object type 'default value'
ok 21 - check addressability for object type 'domain constraint'
ok 22 - check addressability for object type 'event trigger'
ok 23 - check addressability for object type 'extension'
ok 24 - check addressability for object type 'foreign table'
ok 25 - check addressability for object type 'foreign table column'
ok 26 - check addressability for object type 'foreign-data wrapper'
ok 27 - check addressability for object type 'function'
ok 28 - check addressability for object type 'function of access method'
ok 29 - check addressability for object type 'index'
ok 30 - check addressability for object type 'index column'
ok 31 - check addressability for object type 'language'
ok 32 - check addressability for object type 'large object'
ok 33 - check addressability for object type 'materialized view'
ok 34 - check addressability for object type 'materialized view column'
ok 35 - check addressability for object type 'operator'
ok 36 - check addressability for object type 'operator class'
ok 37 - check addressability for object type 'operator family'
ok 38 - check addressability for object type 'operator of access method'
ok 39 - check addressability for object type 'policy'
ok 40 - check addressability for object type 'role'
ok 41 - check addressability for object type 'rule'
ok 42 - check addressability for object type 'schema'
ok 43 - check addressability for object type 'sequence'
ok 44 - check addressability for object type 'sequence column'
ok 45 - check addressability for object type 'server'
ok 46 - check addressability for object type 'table'
ok 47 - check addressability for object type 'table column'
ok 48 - check addressability for object type 'table constraint'
ok 49 - check addressability for object type 'tablespace'
ok 50 - check addressability for object type 'text search configuration'
ok 51 - check addressability for object type 'text search dictionary'
ok 52 - check addressability for object type 'text search parser'
ok 53 - check addressability for object type 'text search template'
ok 54 - check addressability for object type 'toast table'
ok 55 - check addressability for object type 'toast table column'
ok 56 - check addressability for object type 'transform'
ok 57 - check addressability for object type 'trigger'
ok 58 - check addressability for object type 'type'
ok 59 - check addressability for object type 'user mapping'
ok 60 - check addressability for object type 'view'
ok 61 - check addressability for object type 'view column'
ok 62 - lives_ok: SELECT * FROM cat_tools.object__catalog('access method')
ok 63 - lives_ok: SELECT * FROM cat_tools.object__catalog('aggregate')
ok 64 - lives_ok: SELECT * FROM cat_tools.object__catalog('cast')
ok 65 - lives_ok: SELECT * FROM cat_tools.object__catalog('collation')
ok 66 - lives_ok: SELECT * FROM cat_tools.object__catalog('composite type')
ok 67 - lives_ok: SELECT * FROM cat_tools.object__catalog('composite type column')
ok 68 - lives_ok: SELECT * FROM cat_tools.object__catalog('conversion')
ok 69 - lives_ok: SELECT * FROM cat_tools.object__catalog('database')
ok 70 - lives_ok: SELECT * FROM cat_tools.object__catalog('default acl')
ok 71 - lives_ok: SELECT * FROM cat_tools.object__catalog('default value')
ok 72 - lives_ok: SELECT * FROM cat_tools.object__catalog('domain constraint')
ok 73 - lives_ok: SELECT * FROM cat_tools.object__catalog('event trigger')
ok 74 - lives_ok: SELECT * FROM cat_tools.object__catalog('extension')
ok 75 - lives_ok: SELECT * FROM cat_tools.object__catalog('foreign table')
ok 76 - lives_ok: SELECT * FROM cat_tools.object__catalog('foreign table column')
ok 77 - lives_ok: SELECT * FROM cat_tools.object__catalog('foreign-data wrapper')
ok 78 - lives_ok: SELECT * FROM cat_tools.object__catalog('function')
ok 79 - lives_ok: SELECT * FROM cat_tools.object__catalog('function of access method')
ok 80 - lives_ok: SELECT * FROM cat_tools.object__catalog('index')
ok 81 - lives_ok: SELECT * FROM cat_tools.object__catalog('index column')
ok 82 - lives_ok: SELECT * FROM cat_tools.object__catalog('language')
ok 83 - lives_ok: SELECT * FROM cat_tools.object__catalog('large object')
ok 84 - lives_ok: SELECT * FROM cat_tools.object__catalog('materialized view')
ok 85 - lives_ok: SELECT * FROM cat_tools.object__catalog('materialized view column')
ok 86 - lives_ok: SELECT * FROM cat_tools.object__catalog('operator')
ok 87 - lives_ok: SELECT * FROM cat_tools.object__catalog('operator class')
ok 88 - lives_ok: SELECT * FROM cat_tools.object__catalog('operator family')
ok 89 - lives_ok: SELECT * FROM cat_tools.object__catalog('operator of access method')
ok 90 - lives_ok: SELECT * FROM cat_tools.object__catalog('policy')
ok 91 - lives_ok: SELECT * FROM cat_tools.object__catalog('role')
ok 92 - lives_ok: SELECT * FROM cat_tools.object__catalog('rule')
ok 93 - lives_ok: SELECT * FROM cat_tools.object__catalog('schema')
ok 94 - lives_ok: SELECT * FROM cat_tools.object__catalog('sequence')
ok 95 - lives_ok: SELECT * FROM cat_tools.object__catalog('sequence column')
ok 96 - lives_ok: SELECT * FROM cat_tools.object__catalog('server')
ok 97 - lives_ok: SELECT * FROM cat_tools.object__catalog('table')
ok 98 - lives_ok: SELECT * FROM cat_tools.object__catalog('table column')
ok 99 - lives_ok: SELECT * FROM cat_tools.object__catalog('table constraint')
ok 100 - lives_ok: SELECT * FROM cat_tools.object__catalog('tablespace')
ok 101 - lives_ok: SELECT * FROM cat_tools.object__catalog('text search configuration')
ok 102 - lives_ok: SELECT * FROM cat_tools.object__catalog('text search dictionary')
ok 103 - lives_ok: SELECT * FROM cat_tools.object__catalog('text search parser')
ok 104 - lives_ok: SELECT * FROM cat_tools.object__catalog('text search template')
ok 105 - lives_ok: SELECT * FROM cat_tools.object__catalog('toast table')
ok 106 - lives_ok: SELECT * FROM cat_tools.object__catalog('toast table column')
ok 107 - lives_ok: SELECT * FROM cat_tools.object__catalog('transform')
ok 108 - lives_ok: SELECT * FROM cat_tools.object__catalog('trigger')
ok 109 - lives_ok: SELECT * FROM cat_tools.object__catalog('type')
ok 110 - lives_ok: SELECT * FROM cat_tools.object__catalog('user mapping')
ok 111 - lives_ok: SELECT * FROM cat_tools.object__catalog('view')
ok 112 - lives_ok: SELECT * FROM cat_tools.object__catalog('view column')
ok 113 - lives_ok: SELECT * FROM cat_tools.object__reg_type('access method')
ok 114 - lives_ok: SELECT * FROM cat_tools.object__reg_type('aggregate')
ok 115 - lives_ok: SELECT * FROM cat_tools.object__reg_type('cast')
ok 116 - lives_ok: SELECT * FROM cat_tools.object__reg_type('collation')
ok 117 - lives_ok: SELECT * FROM cat_tools.object__reg_type('composite type')
ok 118 - lives_ok: SELECT * FROM cat_tools.object__reg_type('composite type column')
ok 119 - lives_ok: SELECT * FROM cat_tools.object__reg_type('conversion')
ok 120 - lives_ok: SELECT * FROM cat_tools.object__reg_type('database')
ok 121 - lives_ok: SELECT * FROM cat_tools.object__reg_type('default acl')
ok 122 - lives_ok: SELECT * FROM cat_tools.object__reg_type('default value')
ok 123 - lives_ok: SELECT * FROM cat_tools.object__reg_type('domain constraint')
ok 124 - lives_ok: SELECT * FROM cat_tools.object__reg_type('event trigger')
ok 125 - lives_ok: SELECT * FROM cat_tools.object__reg_type('extension')
ok 126 - lives_ok: SELECT * FROM cat_tools.object__reg_type('foreign table')
ok 127 - lives_ok: SELECT * FROM cat_tools.object__reg_type('foreign table column')
ok 128 - lives_ok: SELECT * FROM cat_tools.object__reg_type('foreign-data wrapper')
ok 129 - lives_ok: SELECT * FROM cat_tools.object__reg_type('function')
ok 130 - lives_ok: SELECT * FROM cat_tools.object__reg_type('function of access method')
ok 131 - lives_ok: SELECT * FROM cat_tools.object__reg_type('index')
ok 132 - lives_ok: SELECT * FROM cat_tools.object__reg_type('index column')
ok 133 - lives_ok: SELECT * FROM cat_tools.object__reg_type('language')
ok 134 - lives_ok: SELECT * FROM cat_tools.object__reg_type('large object')
ok 135 - lives_ok: SELECT * FROM cat_tools.object__reg_type('materialized view')
ok 136 - lives_ok: SELECT * FROM cat_tools.object__reg_type('materialized view column')
ok 137 - lives_ok: SELECT * FROM cat_tools.object__reg_type('operator')
ok 138 - lives_ok: SELECT * FROM cat_tools.object__reg_type('operator class')
ok 139 - lives_ok: SELECT * FROM cat_tools.object__reg_type('operator family')
ok 140 - lives_ok: SELECT * FROM cat_tools.object__reg_type('operator of access method')
ok 141 - lives_ok: SELECT * FROM cat_tools.object__reg_type('policy')
ok 142 - lives_ok: SELECT * FROM cat_tools.object__reg_type('role')
ok 143 - lives_ok: SELECT * FROM cat_tools.object__reg_type('rule')
ok 144 - lives_ok: SELECT * FROM cat_tools.object__reg_type('schema')
ok 145 - lives_ok: SELECT * FROM cat_tools.object__reg_type('sequence')
ok 146 - lives_ok: SELECT * FROM cat_tools.object__reg_type('sequence column')
ok 147 - lives_ok: SELECT * FROM cat_tools.object__reg_type('server')
ok 148 - lives_ok: SELECT * FROM cat_tools.object__reg_type('table')
ok 149 - lives_ok: SELECT * FROM cat_tools.object__reg_type('table column')
ok 150 - lives_ok: SELECT * FROM cat_tools.object__reg_type('table constraint')
ok 151 - lives_ok: SELECT * FROM cat_tools.object__reg_type('tablespace')
ok 152 - lives_ok: SELECT * FROM cat_tools.object__reg_type('text search configuration')
ok 153 - lives_ok: SELECT * FROM cat_tools.object__reg_type('text search dictionary')
ok 154 - lives_ok: SELECT * FROM cat_tools.object__reg_type('text search parser')
ok 155 - lives_ok: SELECT * FROM cat_tools.object__reg_type('text search template')
ok 156 - lives_ok: SELECT * FROM cat_tools.object__reg_type('toast table')
ok 157 - lives_ok: SELECT * FROM cat_tools.object__reg_type('toast table column')
ok 158 - lives_ok: SELECT * FROM cat_tools.object__reg_type('transform')
ok 159 - lives_ok: SELECT * FROM cat_tools.object__reg_type('trigger')
ok 160 - lives_ok: SELECT * FROM cat_tools.object__reg_type('type')
ok 161 - lives_ok: SELECT * FROM cat_tools.object__reg_type('user mapping')
ok 162 - lives_ok: SELECT * FROM cat_tools.object__reg_type('view')
ok 163 - lives_ok: SELECT * FROM cat_tools.object__reg_type('view column')
ok 164 - Verify cat_tools.object__address_classid('access method')
ok 165 - Verify cat_tools.object__address_classid('aggregate')
ok 166 - Verify cat_tools.object__address_classid('cast')
ok 167 - Verify cat_tools.object__address_classid('collation')
ok 168 - Verify cat_tools.object__address_classid('composite type')
ok 169 - Verify cat_tools.object__address_classid('composite type column')
ok 170 - Verify cat_tools.object__address_classid('conversion')
ok 171 - Verify cat_tools.object__address_classid('database')
ok 172 - Verify cat_tools.object__address_classid('default acl')
ok 173 - Verify cat_tools.object__address_classid('default value')
ok 174 - Verify cat_tools.object__address_classid('domain constraint')
ok 175 - Verify cat_tools.object__address_classid('event trigger')
ok 176 - Verify cat_tools.object__address_classid('extension')
ok 177 - Verify cat_tools.object__address_classid('foreign table')
ok 178 - Verify cat_tools.object__address_classid('foreign table column')
ok 179 - Verify cat_tools.object__address_classid('foreign-data wrapper')
ok 180 - Verify cat_tools.object__address_classid('function')
ok 181 - Verify cat_tools.object__address_classid('function of access method')
ok 182 - Verify cat_tools.object__address_classid('index')
ok 183 - Verify cat_tools.object__address_classid('index column')
ok 184 - Verify cat_tools.object__address_classid('language')
ok 185 - Verify cat_tools.object__address_classid('large object')
ok 186 - Verify cat_tools.object__address_classid('materialized view')
ok 187 - Verify cat_tools.object__address_classid('materialized view column')
ok 188 - Verify cat_tools.object__address_classid('operator')
ok 189 - Verify cat_tools.object__address_classid('operator class')
ok 190 - Verify cat_tools.object__address_classid('operator family')
ok 191 - Verify cat_tools.object__address_classid('operator of access method')
ok 192 - Verify cat_tools.object__address_classid('policy')
ok 193 - Verify cat_tools.object__address_classid('role')
ok 194 - Verify cat_tools.object__address_classid('rule')
ok 195 - Verify cat_tools.object__address_classid('schema')
ok 196 - Verify cat_tools.object__address_classid('sequence')
ok 197 - Verify cat_tools.object__address_classid('sequence column')
ok 198 - Verify cat_tools.object__address_classid('server')
ok 199 - Verify cat_tools.object__address_classid('table')
ok 200 - Verify cat_tools.object__address_classid('table column')
ok 201 - Verify cat_tools.object__address_classid('table constraint')
ok 202 - Verify cat_tools.object__address_classid('tablespace')
ok 203 - Verify cat_tools.object__address_classid('text search configuration')
ok 204 - Verify cat_tools.object__address_classid('text search dictionary')
ok 205 - Verify cat_tools.object__address_classid('text search parser')
ok 206 - Verify cat_tools.object__address_classid('text search template')
ok 207 - Verify cat_tools.object__address_classid('toast table')
ok 208 - Verify cat_tools.object__address_classid('toast table column')
ok 209 - Verify cat_tools.object__address_classid('transform')
ok 210 - Verify cat_tools.object__address_classid('trigger')
ok 211 - Verify cat_tools.object__address_classid('type')
ok 212 - Verify cat_tools.object__address_classid('user mapping')
ok 213 - Verify cat_tools.object__address_classid('view')
ok 214 - Verify cat_tools.object__address_classid('view column')
ok 215 - Change search_path
ok 216 - Create bogus pg_class table
ok 217 - Create bogus regclass type
ok 218 - Simple 'pg_class'::pg_catalog.regclass should not return pg_catalog.pg_class
ok 219 - Simple 'regclass'::regtype should not return pg_catalog.regtype
ok 220 - cat_tools.object__catalog('table') returns pg_catalog.pg_class
ok 221 - cat_tools.object__catalog('table') returns pg_catalog.pg_class
ok 222 - Verify objects__shared_src() returns correct values
# TRANSACTION INTENTIONALLY LEFT OPEN!
